internal/transport/transport_test.go

```diff
@@ -2664,3 +2664,91 @@ func TestConnectionError_Unwrap(t *testing.T) {
 		t.Error("ConnectionError does not unwrap")
 	}
 }
+
+// Test that a transport level client shutdown successfully sends a GOAWAY frame
+// to underlying connection.
+func (s) TestClientSendsAGoAwayFrame(t *testing.T) {
+	// Create a server.
+	lis, err := net.Listen("tcp", "localhost:0")
+	if err != nil {
+		t.Fatalf("Error while listening: %v", err)
+	}
+	defer lis.Close()
+	// The success channel verifies that the server's reader goroutine received
+	// a GOAWAY frame from the client.
+	greetDone := make(chan struct{})
+	successCh := make(chan struct{})
+	errorCh := make(chan struct{})
+	connectCtx, cancel := context.WithDeadline(context.Background(), time.Now().Add(10*time.Second))
+	defer cancel()
+	// Launch the server.
+	go func() {
+		sconn, err := lis.Accept()
+		if err != nil {
+			t.Errorf("Error while accepting: %v", err)
+		}
+		defer func(sconn net.Conn) {
+			err := sconn.Close()
+			if err != nil {
+				return
+			}
+		}(sconn)
+		if _, err := io.ReadFull(sconn, make([]byte, len(clientPreface))); err != nil {
+			t.Errorf("Error while writing settings ack: %v", err)
+			return
+		}
+		sfr := http2.NewFramer(sconn, sconn)
+		if err := sfr.WriteSettings(); err != nil {
+			t.Errorf("Error while writing settings %v", err)
+			return
+		}
+
+		sfr.ReadFrame()
+		sfr.ReadFrame()
+		sfr.ReadFrame()
+		close(greetDone)
+
+		frame, err := sfr.ReadFrame()
+		if err != nil {
+			return
+		}
+		switch fr := frame.(type) {
+		case *http2.GoAwayFrame:
+			// Records that the server successfully received a GOAWAY frame.
+			goAwayFrame := fr
+			if goAwayFrame.ErrCode == http2.ErrCodeNo && string(goAwayFrame.DebugData()) == "graceful_shutdown" {
+				t.Logf("Received goAway frame from client")
+				close(successCh)
+			} else {
+				close(errorCh)
+			}
+			return
+		default:
+			// The client should have sent any frame other than GOAWAY
+			t.Logf("The server received a frame other than GOAWAY")
```

Comment:
Seems like this is also a failure path? We should `close(errorCh)` here as well?